app/models/ems_refresh/save_inventory.rb

```diff
@@ -181,6 +181,29 @@ def save_vms_inventory(ems, hashes, target = nil)
     end
   end
 
+  # Convert all mapped hashes into actual tags and associate them with the object.
+  # The hash[:tags] object should be an array of hashes, probably created by
+  # the ContainerLabelTagMapping.map_labels method. Each hash in the array should
+  # have the following basic structure:
+  #
+  # {:category_tag_id=>139, :entry_name=>"foo", :entry_description=>"bar"}
+  #
+  # For now, the object must be a VmOrTemplate object.
+  #
+  def save_tags_inventory(object, hash)
+    return if hash.blank?
+    return unless object.kind_of?(VmOrTemplate)
+
+    tags = hash[:tags].collect do |tag_hash|
+      ContainerLabelTagMapping.find_or_create_tag(tag_hash)
```

Comment:
Ah, didn't notice VmOrTemplate.  But IIUC you can't really have 2 methods with same name, only one wins (probably the save_containers one without restriction, or tests would fail).
Why do you need this restriction?  You control which models get tagged by choosing where to calls to `save_tags_inventory` (typically via :tags in child keys lists).  Also, in what situation would you expect to have computed `map_labels`, only to throw it away here?